 # Importing needed packages

In [ ]:
import numpy as np 
import pandas as pd
from sklearn import preprocessing
from sklearn.metrics import confusion_matrix , classification_report  , accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics


# Reading the data in pandas and preview it

In [45]:
df=pd.read_csv("F:\DATA\MAIN\/Heart Attack Risk.csv")
print(df.head)
print(df.value_counts)


<bound method NDFrame.head of      age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  outp

# Targeting x , y 


In [46]:
x=df.drop(['output'] ,  axis= 1 )
y=df['output']


# test and train split (manual)

In [47]:

msk=np.random.rand(len(df)) < 0.8

train_x = x[msk]
test_x = x[~msk]
train_y = y[msk]
test_y = y[~msk]         


# modeling 

# checking the resault with LogisticRegression

In [ ]:
from sklearn.linear_model import LogisticRegression

LR=LogisticRegression(C=1 , solver='liblinear').fit(train_x , train_y)
solvers = ['liblinear', 'sag', 'saga', 'lbfgs', 'newton-cholesky', 'newton-cg']
best_solver = ''
best_accuracy = 0

for solver in solvers:
        model = LogisticRegression(solver=solver, max_iter=1000).fit(train_x, train_y)
        test_y_ = model.predict(test_x)
        accuracy = accuracy_score(test_y, test_y_)
        
        print(f"Solver: {solver}, Accuracy: {accuracy}")
        
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_solver = solver


print(f"Best Solver: {best_solver}, Best Accuracy: {best_accuracy}")

conf_matrix = confusion_matrix(test_y, test_y_)
print("Confusion Matrix:", conf_matrix)



Solver: liblinear, Accuracy: 0.8524590163934426
Solver: sag, Accuracy: 0.7213114754098361
Solver: saga, Accuracy: 0.6721311475409836


c:\Users\korosh king\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\korosh king\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


Solver: lbfgs, Accuracy: 0.8524590163934426
Solver: newton-cholesky, Accuracy: 0.8524590163934426
Solver: newton-cg, Accuracy: 0.8524590163934426
Best Solver: liblinear, Best Accuracy: 0.8524590163934426
Confusion Matrix:
 [[23  5]
 [ 4 29]]


 -the accuracy of LR is very good and the FN is 5 which is acceptable

# checking the resault with SVM

In [66]:
from sklearn.svm import SVC

# لیست کرنل‌ها
kernels = ['sigmoid', 'poly', 'rbf', 'precomputed', 'linear']
best_kernel = ''
best_accuracy = 0

# حلقه برای تست کرنل‌ها
for kernel in kernels:
    try:
        model = SVC(kernel=kernel, max_iter=1000).fit(train_x, train_y)
        test_y_ = model.predict(test_x)
        accuracy = accuracy_score(test_y, test_y_)
        
        # چاپ دقت برای هر کرنل
        print(f"Kernel: {kernel}, Accuracy: {accuracy}")
        
        # به‌روزرسانی بهترین کرنل و دقت
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_kernel = kernel
    except:
        print(f"Kernel: {kernel} failed")

print(f"Best Kernel: {best_kernel}, Best Accuracy: {best_accuracy}")


conf_matrix = confusion_matrix(test_y, test_y_)
print("Confusion Matrix:", conf_matrix)


Kernel: sigmoid, Accuracy: 0.5245901639344263
Kernel: poly, Accuracy: 0.639344262295082
Kernel: rbf, Accuracy: 0.6229508196721312
Kernel: precomputed failed
Kernel: linear, Accuracy: 0.6065573770491803
Best Kernel: poly, Best Accuracy: 0.639344262295082
Confusion Matrix: [[13 15]
 [ 9 24]]


c:\Users\korosh king\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\svm\_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


-so we undrestand that SVM is not a proper model for this data and also FN is to high 

# Now Lets try the resault with DecisionTreeClassifier

In [ ]:
from sklearn.tree import DecisionTreeClassifier

criteria = ['gini', 'entropy', 'log_loss']
best_criterion = ''
best_accuracy = 0

for criterion in criteria:
    try:
        dTree = DecisionTreeClassifier(criterion=criterion, max_depth=4).fit(train_x, train_y)
        test_y_ = dTree.predict(test_x)
        accuracy = accuracy_score(test_y, test_y_)
        
        print(f"Criterion: {criterion}, Accuracy: {accuracy}")
        
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_criterion = criterion
    except:
        print(f"Criterion: {criterion} failed")

print(f"Best Criterion: {best_criterion}, Best Accuracy: {best_accuracy}")
conf_matrix = confusion_matrix(test_y, test_y_)
print("Confusion Matrix:", conf_matrix)


Criterion: gini, Accuracy: 0.7704918032786885
Criterion: entropy, Accuracy: 0.819672131147541
Criterion: log_loss, Accuracy: 0.819672131147541

Best Criterion: entropy, Best Accuracy: 0.819672131147541
Confusion Matrix: [[23  5]
 [ 6 27]]


- - ALso the accuracy of DTree is very good and the FN is 5 again which is acceptable 

# Based on this program, we found that the best classifier models are Decision Tree (DTree) and Logistic Regression (LR). Additionally, by adjusting the parameters, I realized that the best criterion for DTree is entropy, and for Logistic Regression, the best solver is liblinear, both achieving approximately 85% accuracy.But SVM is not a correct model for this data frame :)) .
